lib/Backends/OpenCL/OpenCL.cpp

```diff
@@ -1137,23 +1149,63 @@ Error OpenCLFunction::execute(ExecutionContext *context) {
 
       // This is a naive implementation that parallelizes using three dims:
       // the X and the Y in the output filter.
-      cl_kernel kernel = createKernel(kernelName, program);
-      setKernelArg(kernel, 0, deviceBuffer);
-      auto numArgs = setKernelArgsForBuffers(kernel, I, 1, runtimeBundle_);
-      auto odim = ShapeNHWC(CC->getDest()->getType()->dims());
+
+      // Specialize the kernel related to parameters with very frequent
+      // values to enable constant propagation and other aggressive
+      // optimizations.
       auto idim = ShapeNHWC(CC->getSrc()->getType()->dims());
-      auto pads = PaddingTLBR(CC->getPads());
+
+      std::string src;
+      src.append(reinterpret_cast<const char *>(kernels_cl_src),
+                 kernels_cl_src_size);
+
       ShapeHW kdim(CC->getKernels());
       ShapeHW sdim(CC->getStrides());
+      auto odim = ShapeNHWC(CC->getDest()->getType()->dims());
+      ShapeNHWC kernelSize(CC->getFilter()->getType()->dims());
+      auto pads = PaddingTLBR(CC->getPads());
+
+      std::vector<std::string> options;
+
+      if (clSpecializeConvolutionK) {
+        // Specialize convolutionK aggressively, similar to conv_forward_mem
+        if (CC->getGroup() == 1)
+          addIntOption(options, "CONVK_GROUP", 1);
+        if (idim.n == 1)
+          addIntOption(options, "CONVK_BATCHES", 1);
+        addIntOption(options, "CONVK_DILATION", CC->getDilation());
+        addIntOption(options, "CONVK_KERNEL_W", kdim.width);
+        addIntOption(options, "CONVK_KERNEL_H", kdim.height);
+        addIntOption(options, "CONVK_STRIDES_W", sdim.width);
+        addIntOption(options, "CONVK_STRIDES_H", sdim.height);
+        addIntOption(options, "CONVK_IDIM_W", idim.w);
+        addIntOption(options, "CONVK_IDIM_H", idim.h);
+        addIntOption(options, "CONVK_IDIM_C", idim.c);
+        addIntOption(options, "CONVK_ODIM_W", odim.w);
+        addIntOption(options, "CONVK_ODIM_H", odim.h);
+        addIntOption(options, "CONVK_ODIM_C", odim.c);
+        addIntOption(options, "CONVK_PADS_TOP", pads.top);
+        addIntOption(options, "CONVK_PADS_LEFT", pads.left);
+        addIntOption(options, "CONVK_FILTER_W", kernelSize.w);
+        addIntOption(options, "CONVK_FILTER_H", kernelSize.h);
+        addIntOption(options, "CONVK_FILTER_C", kernelSize.c);
+        addIntOption(options, "CONVK_SPECIALIZED", 1);
+      }
+
+      cl_program specializedProg = createProgram(src, options, commands);
```

Comment:
This will parse/compile the whole set of kernels, even though you try to specialize only one of them, right?

Reply:
Yes, whatever is in that source string gets compiled, so this could be sped up somewhat if the specialized convolutionK was isolated to a separate file. I think the impact of this would be very minor as the optimizations (that are often performed only for the built kernel) take most of the compile time.

In any case, the compile time impact of the specialization is quite high for the first run anyways (and should be considered/discarded as initialization time) so I wouldn't do this solely for compile time, but possibly it would be cleaner if the specialized convolutionK was in another file.